strategies/categorized/src/main/java/org/arquillian/smart/testing/strategies/categorized/CategoriesParser.java

```diff
@@ -0,0 +1,88 @@
+package org.arquillian.smart.testing.strategies.categorized;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.arquillian.smart.testing.logger.Log;
+import org.arquillian.smart.testing.logger.Logger;
+
+class CategoriesParser {
+
+    private final Logger logger = Log.getLogger();
+    private final CategorizedConfiguration strategyConfig;
+    private final List<String> specifiedCategories;
+
+    CategoriesParser(CategorizedConfiguration strategyConfig) {
+        this.strategyConfig = strategyConfig;
+        specifiedCategories = Arrays.stream(strategyConfig.getCategories())
+            .map(this::changeIfNonCaseSensitive)
+            .collect(Collectors.toList());
+    }
+
+    boolean hasCorrectCategories(Class<?> clazz) {
+        List<String> presentCategories = getPresentCategories(clazz);
+
+        if (!presentCategories.isEmpty()) {
+            if (specifiedCategories.size() > 0) {
+
+                List<String> intersection = presentCategories.stream()
+                    .filter(category -> isSpecified(category, specifiedCategories))
+                    .collect(Collectors.toList());
+
+                if (strategyConfig.isMatchAll()) {
+                    return intersection.size() == specifiedCategories.size();
+                } else {
+                    return !intersection.isEmpty();
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private List<String> getPresentCategories(Class<?> clazz) {
+        return Arrays.stream(clazz.getAnnotations())
+            .filter(this::isJUnit4Category)
+            .flatMap(this::retrieveCategoriesFromAnnotation)
+            .map(this::changeIfNonCaseSensitive)
+            .collect(Collectors.toList());
+    }
+
+    private boolean isSpecified(String category, List<String> specifiedCategories) {
+        if (!specifiedCategories.contains(category)) {
+            if (category.contains(".")) {
+                String categoryClassName = category.substring(category.lastIndexOf(".") + 1);
+                return specifiedCategories.contains(categoryClassName);
+            }
+            return false;
+        }
+        return true;
+    }
+
+    private String changeIfNonCaseSensitive(String category) {
+        return strategyConfig.isCaseSensitive() ? category : category.toLowerCase();
+    }
+
+    private Stream<String> retrieveCategoriesFromAnnotation(Annotation categoryAnnotation) {
+        try {
+            Method valueMethod = categoryAnnotation.getClass().getMethod("value");
+            return Arrays.asList((Class[]) valueMethod.invoke(categoryAnnotation))
```

Comment:
This can simply be replaced with `Arrays.stream()`

Reply:
Yeah, right... zombie :-)